### PR TITLE
Push Characters, Character Attributes and Asset Types Controllers (C,R,U,D)

### PR DESCRIPTION
**Pull request by AnhQuoc2906**

### CODE REVIEW

File: Api/Controllers/AssetTypesController.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+using ServiceLayer.Business;
+using WebApiLayer.UserFeatures.Requests;
+
+namespace WebApiLayer.Controllers;
+
+[Route(Constants.HTTP.API_VERSION + "/gms/asset-types")]
+public class AssetTypesController : BaseController
+{
+    private readonly IAssetTypeServices _assetTypeServices;
+    private readonly IGenericRepository<AssetTypeEntity> _assetTypeRepo;
```
Comment (danielNg91) [inline Api/Controllers/AssetTypesController.cs:14]: this field is unused, remove it

File: Api/Controllers/CharactersController.cs [hunk 2]
```diff
@@ -1,35 +1,58 @@
...
 
 [Route(Constants.HTTP.API_VERSION + "/gms/characters")]
 public class CharactersController : BaseController
 {
+    private readonly ICharacterServices _characterServices;
+    private readonly IGenericRepository<CharacterEntity> _characterRepo;
+    public CharactersController(ICharacterServices characterServices, IGenericRepository<CharacterEntity> characterRepo)
+    {
+        _characterServices = characterServices;
+        _characterRepo = characterRepo;
```
Comment (danielNg91) [inline Api/Controllers/CharactersController.cs:18]: and this

**Review by danielNg91 — APPROVED** [review]
fix corresponding to my comment, then you can merge it

File: Application/Business/LevelServices/LevelServices.cs [hunk 3]
```diff
@@ -62,8 +54,16 @@ public async Task CheckLevel(LevelEntity level)
         {
             if(level.Id == Guid.Empty || levelCheck.Id != level.Id)
             {
-                throw new BadRequestException("The game already has this level's name");
+                throw new BadRequestException("The user already have a character in this game server");
             }
         }
     }
+    public async Task CheckLevel(Guid id)
```
Comment (danielNg91) [inline Application/Business/LevelServices/LevelServices.cs:61]: so many duplicated logics, how about create a function FoundOrThrow(...args) in repo and reuse it?

**Review by danielNg91 — CHANGES_REQUESTED** [review]

File: Application/Business/LevelServices/LevelServices.cs (hunk 3 above)
Comment (AnhQuoc2906) [reply Application/Business/LevelServices/LevelServices.cs:61]: Add new function in Generic Repository:
```
    public virtual async Task CheckExistAsync(Guid id, string message, params string[] navigationProperties)
    {
        var query = ApplyNavigation(navigationProperties);
        T entity = await query.FirstOrDefaultAsync(e => e.Id.Equals(id));
        if(entity is null)
        {
            throw new NotFoundException(message);
        }
    }
```
Remove all Check... functions in services

File: Application/Business/Character/CharacterServices.cs [hunk 4]
```diff
@@ -37,13 +38,30 @@ public async Task<int> Count()
...
     public async Task Create(CharacterEntity character)
     {
+        await CheckCharacter(character);
+        await _characterRepo.CreateAsync(character);
     }
     public async Task Update(Guid characterId, CharacterEntity character)
     {
-
+        await _characterRepo.CheckExistAsync(characterId, "Character not exist.");
+        await CheckCharacter(character);
```
Comment (danielNg91) [inline Application/Business/Character/CharacterServices.cs:47]: What to check? function name should be describe what it do

**Review by danielNg91 — CHANGES_REQUESTED** [review]

File: Api/Controllers/CharactersController.cs [hunk 5]
```diff
@@ -1,35 +1,56 @@
...
+        Mapper.Map(character, newC);
+        await _characterServices.Create(newC);
+        return CreatedAtAction("GetCharacter", new { id = newC.Id }, newC);
     }
 
     [HttpPut("{id}")]
-    public void Put(int id, [FromBody] string value)
+    public async Task<IActionResult> UpdateCharacter(Guid id, [FromBody] UpdateCharacterRequest character)
     {
+        var newC = await _characterServices.GetById(id);
```
Comment (danielNg91) [inline Api/Controllers/CharactersController.cs:44]: Find the entity in the repo here, why not throw it when not found but wait for the service to throw it??

**Review by danielNg91 — CHANGES_REQUESTED** [review]

File: Application/Business/AssetType/AssetTypeServices.cs [hunk 6]
```diff
@@ -30,14 +31,30 @@ public async Task<int> Count()
...
     {
-
+        await _assetTypeRepo.FoundOrThrowAsync(assetTypeId, "Asset type is not exist.");
+        await _assetTypeRepo.DeleteSoftAsync(assetTypeId);
+    }
+    public async Task CheckForDuplicateAssetType(AssetTypeEntity assetType)
+    {
+        var checkAssetType = await _assetTypeRepo.FirstOrDefaultAsync(
+            aT => aT.Name.Equals(assetType.Name) && aT.GameId.Equals(assetType.GameId));
+        if(checkAssetType is not null)
```
Comment (danielNg91) [inline Application/Business/AssetType/AssetTypeServices.cs:52]: this nested if can reduce by using where clause and check list.Count > 1.
FirstOrDefault returns the first element that satisfies the query, but what if you always get the element that satisfies the if clause but in the database there exist other elements that do not satisfy?

File: Application/Business/AttributeGroup/AttributeGroupServices.cs [hunk 7]
```diff
@@ -32,21 +32,12 @@ public async Task<int> Count()
     }
     public async Task Update(Guid attributeGroupid, AttributeGroupEntity attributeGroup)
     {
-        var target = await GetById(attributeGroupid);
-        if(target is null)
-        {
-            throw new NotFoundException("Attribute group not exist");
-        }
-        await _attributeRepo.UpdateAsync(target);
+        await _attributeRepo.FoundOrThrowAsync(attributeGroupid, "Attribute group not exist.");
```
Comment (danielNg91) [inline Application/Business/AttributeGroup/AttributeGroupServices.cs:35]: rewrite here too

File: Application/Business/Character/CharacterServices.cs [hunk 8]
```diff
@@ -37,13 +38,30 @@ public async Task<int> Count()
     }
     public async Task Create(CharacterEntity character)
     {
+        await CheckForDuplicateCharacter(character);
+        await _characterRepo.CreateAsync(character);
     }
     public async Task Update(Guid characterId, CharacterEntity character)
     {
-
+        await _characterRepo.FoundOrThrowAsync(characterId, "Character not exist.");
```
Comment (danielNg91) [inline Application/Business/Character/CharacterServices.cs:46]: too

File: Application/Business/Character/CharacterServices.cs [hunk 9]
```diff
@@ -37,13 +38,30 @@ public async Task<int> Count()
...
+        await CheckForDuplicateCharacter(character);
+        await _characterRepo.UpdateAsync(character);
     }
     public async Task Delete(Guid characterId)
     {
-
+        await _characterRepo.FoundOrThrowAsync(characterId, "Character not exist.");
+        await _characterRepo.DeleteSoftAsync(characterId);
+    }
+    public async Task CheckForDuplicateCharacter(CharacterEntity character)
```
Comment (danielNg91) [inline Application/Business/Character/CharacterServices.cs:55]: too

File: Application/Business/CharacterAttribute/CharacterAttributeServices.cs [hunk 10]
```diff
@@ -33,14 +34,23 @@ public async Task<int> Count()
     }
     public async Task Create(CharacterAttributeEntity characterAttribute)
     {
-
+        var charAttCheck = await _characterAttributeRepo.FirstOrDefaultAsync(
```
Comment (danielNg91) [inline Application/Business/CharacterAttribute/CharacterAttributeServices.cs:37]: why we need this double 2 kind of id?

File: Application/Business/CharacterAttribute/CharacterAttributeServices.cs [hunk 11]
```diff
@@ -33,14 +34,23 @@ public async Task<int> Count()
...
+        if(charAttCheck is not null)
+        {
+            throw new BadRequestException("Character with this attribute group already exist.");
+        }
+        await _characterAttributeRepo.CreateAsync(characterAttribute);
     }
     public async Task Update(Guid characterAttributeid, CharacterAttributeEntity characterAttribute)
     {
-
+        await _characterAttributeRepo.FoundOrThrowAsync(characterAttributeid, "Character attribute not exist.");
```
Comment (danielNg91) [inline Application/Business/CharacterAttribute/CharacterAttributeServices.cs:47]: rewrite too

File: Application/Business/CharacterType/CharacterTypeServices.cs [hunk 12]
```diff
@@ -42,22 +42,12 @@ public async Task Create(CharacterTypeEntity characterType)
     }
     public async Task Update(Guid characterTypeId, CharacterTypeEntity characterType)
     {
-        var ctCheck = await _characterTypeRepo.FindByIdAsync(characterTypeId);
-        if (ctCheck is null)
-        {
-            throw new BadRequestException("Character type not exist");
-        }
+        await _characterTypeRepo.FoundOrThrowAsync(characterTypeId, "Character type not exist.");
```
Comment (danielNg91) [inline Application/Business/CharacterType/CharacterTypeServices.cs:45]: too

File: Application/Business/GameServer/GameServerServices.cs [hunk 13]
```diff
@@ -35,20 +35,13 @@ public async Task Create(GameServerEntity gameServer)
     }
     public async Task Update(Guid gameServerId, GameServerEntity gameServer)
     {
-        var target = await GetById(gameServerId);
-        if (target is null)
-        {
-            throw new NotFoundException("Game server not exist");
-        }
-        await _gameServerRepo.UpdateAsync(target);
+        await _gameServerRepo.FoundOrThrowAsync(gameServerId, "Game server not exist.");
```
Comment (danielNg91) [inline Application/Business/GameServer/GameServerServices.cs:38]: too

File: Application/Business/GameServices/GameServices.cs [hunk 14]
```diff
@@ -42,21 +42,12 @@ public async Task Create(GameEntity game)
     }
     public async Task Update(Guid gameId, GameEntity game)
     {
-        var gameCheck = await _gameRepo.FirstOrDefaultAsync(
-            g => g.Id.Equals(gameId));
-        if (gameCheck == null)
-        {
-            throw new BadRequestException("Game not exist");
-        }
+        await _gameRepo.FoundOrThrowAsync(gameId, "Game not exist.");
```
Comment (danielNg91) [inline Application/Business/GameServices/GameServices.cs:45]: too

File: Application/Business/LevelServices/LevelServices.cs [hunk 15]
```diff
@@ -35,34 +35,26 @@ public async Task<int> Count(Guid gameId)
...
-        if(level is null)
-        {
-            throw new BadRequestException("Level not exist");
-        }
-        await _levelRepo.DeleteSoftAsync(level);
+        await _levelRepo.FoundOrThrowAsync(levelId, "Level not exist.");
+        await _levelRepo.DeleteSoftAsync(levelId);
     }
-    public async Task CheckLevel(LevelEntity level)
+    public async Task CheckForDuplicateLevel(LevelEntity level)
```
Comment (danielNg91) [inline Application/Business/LevelServices/LevelServices.cs:50]: t

File: Application/Business/UserServices/UserServices.cs [hunk 16]
```diff
@@ -32,24 +31,16 @@ public async Task<int> Count()
...
         await _userRepo.CreateAsync(user);
     }
     public async Task Update(Guid UserId, UserEntity user) {
-        var target = await GetById(UserId);
-        if(target is null)
-        {
-            throw new NotFoundException("User not exist");
-        }
-        await _userRepo.UpdateAsync(target);
+        await _userRepo.FoundOrThrowAsync(UserId, "User not exist.");
```
Comment (danielNg91) [inline Application/Business/UserServices/UserServices.cs:39]: too

File: Infrastructure/Repositories/GenericRepository.cs [hunk 17]
```diff
@@ -75,6 +73,21 @@ public virtual async Task<T> FindByIdAsync(Guid id, params string[] navigationPr
         return entity;
     }
 
+    public virtual async Task<T> FoundOrThrowAsync(Guid id, string message, params string[] navigationProperties)
+    {
+        var query = ApplyNavigation(navigationProperties);
+        T entity = await query.FirstOrDefaultAsync(e => e.Id.Equals(id));
+        if (entity is null)
```
Comment (danielNg91) [inline Infrastructure/Repositories/GenericRepository.cs:80]: use [Optional Argument](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments?redirectedfrom=MSDN) like this

File: Infrastructure/Repositories/GenericRepository.cs [hunk 18]
```diff
@@ -75,6 +73,21 @@ public virtual async Task<T> FindByIdAsync(Guid id, params string[] navigationPr
...
+        {
+            if (string.IsNullOrEmpty(message))
```
Comment (danielNg91) [inline Infrastructure/Repositories/GenericRepository.cs:82]: use [Optional Argument](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments?redirectedfrom=MSDN) like this

File: Infrastructure/Repositories/IGenericRepository.cs [hunk 19]
```diff
@@ -9,6 +11,7 @@ public interface IGenericRepository<T>
     Task CreateRangeAsync(IEnumerable<T> entities);
     Task<List<T>> ListAsync();
     Task<T> FindByIdAsync(Guid id, params string[] navigationProperties);
+    Task<T> FoundOrThrowAsync(Guid id, string message = Constants.ERROR.NOT_FOUND_ERROR, params string[] navigationProperties);
```
Comment (danielNg91) [inline Infrastructure/Repositories/IGenericRepository.cs:14]: we don't need add default value in interface

**Review by danielNg91 — CHANGES_REQUESTED** [review]

File: Api/Controllers/AssetTypesController.cs [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+        var uAssetType = await _assetTypeRepo.FoundOrThrowAsync(id, "Asset Type not exist.");
+        Mapper.Map(assetType, uAssetType);
+        await _assetTypeServices.Update(id, uAssetType);
+        return Ok(uAssetType);
+    }
+
+    [HttpDelete("{id}")]
+    public async Task<IActionResult> DeleteAssetType(Guid id)
+    {
+        var uAssetType = await _assetTypeRepo.FoundOrThrowAsync(id, "Asset Type not exist.");
```
Comment (danielNg91) [inline Api/Controllers/AssetTypesController.cs:56]: this is unsed var

**Review by danielNg91 — CHANGES_REQUESTED** [review]

**Review by danielNg91 — APPROVED** [review]

**Pull request merged**